Support function registration with package-url

rdhabalia opened the pull request:
### Motivation

Right now, function registration uploads function-pkg to bookkeeper and downloads it from bk. However, in some cases, user might already have  func-pkg available at some location and there should be a way to provide a url-location of func-pkg from which pulsar-func-service downloads pkg-file and loads it.

### Modifications

- add support to register function by just giving url-location of func-pkg.
Right now, it supports http/https/file protocols in url.

### Result

user can register function by just providing url-location of the package.

rdhabalia commented:
@srkukarni

sijie commented:
just one general comment, do we really need a separate endpoint for this? Can't we extend existing registerFunction and updateFunction to handle url? filePath is a special url, right?

jerrypeng commented on pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/FunctionsImpl.java:
```diff
@@ -121,6 +121,23 @@ public void createFunction(FunctionDetails functionDetails, String fileName) thr
     }
 
+    @Override
+    public void createFunctionWithUrl(FunctionDetails functionDetails, String pkgUrl) throws PulsarAdminException {
```
Is this used somewhere? Don't we need to modify the CmdFunctions to use this?

rdhabalia replied:
this feature will be used by #1904 . yes, I was thinking to merge this new API change and then I can make appropriate changes at its usage .

rdhabalia commented:
@sijie 
> just one general comment, do we really need a separate endpoint for this? 

right now, `register-function` mainly takes function-metadata and file-stream as argument. file-stream  and file-url are mutually exclusive so, I felt to create a new resource-endpoint where url can replace the file-stream.
However, you are right, instead of a new end-point, I think we can add url along with file-stream in the same register-function api and user can provide either of them and we can document it as well.

jerrypeng commented on pulsar-functions/proto/src/main/proto/Function.proto:
```diff
@@ -85,6 +85,7 @@ message SinkSpec {
 
 message PackageLocationMetaData {
     string packagePath = 1;
+    string packageUrl = 2;
```
should we have another field here? or can we just use the existing field 'packagePath' and switch logic on the prefix e.g. http

rdhabalia replied:
I wanted to avoid deriving logic which can create issue if url and path has conflicting convention in future. however, as @sijie mentioned if we want to deprecate jar-submission then we can just keep one field. fixed it.

sijie commented:
> However, you are right, instead of a new end-point, I think we can add url along with file-stream in the same register-function api and user can provide either of them and we can document it as well.

Yeah.

I think we can deprecate 'filePath' and encourage to use 'fileUrl'. because `fileUrl` is more generic than `filePath`. keeping `filePath` is just for BC consideration.

rdhabalia commented:
> I  think we can deprecate 'filePath' and encourage to use 'fileUrl'. because fileUrl is more generic than filePath. keeping filePath is just for BC consideration.

Sure, I have removed separate admin-api for file-url and added to the same existing api.

@jerrypeng addressed comments.

rdhabalia commented:
@sijie addressed comments.

rdhabalia commented:
@sijie @jerrypeng any more comments? can we merge this one?

saandrews commented on pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionActioner.java:
```diff
@@ -103,18 +105,44 @@ public void join() throws InterruptedException {
+        if(isPkgUrlProvided && pkgLocation.startsWith(Utils.FILE)) {
+            pkgFile = new File(pkgLocation);
+        } else {
+            downloadFile(pkgFile, isPkgUrlProvided, functionMetaData, instanceId);
```
Is it retried when downloadFile comes back with exception?

saandrews replied:
Thats good. How do we identify cases where it fails permanently(e.g: file not in the location)?
